### PR TITLE
Support for `index_addr` instruction

### LINKED ISSUE #77: Support `index_addr` instruction

### PR DESCRIPTION
**Pull request by taufique71**
Added WALA support for `index_addr` instruction

Resolves [Issue 77](https://github.com/themaplelab/swift/issues/77).

### CODE REVIEW

**Comment by taufique71** [comment]
@mmroz Can you please review the changes and point me out to things I am doing wrong or missing?

File: include/swift/WALASupport/SILWalaInstructionVisitor.h [hunk 1]
```diff
@@ -12,8 +12,7 @@ namespace swift {
 class SILWalaInstructionVisitor : public SILInstructionVisitor<SILWalaInstructionVisitor, jobject> {
 public:
   SILWalaInstructionVisitor(const WALAIntegration &Wala, bool Print) : Print(Print), Wala(Wala) {}
-
-  void visitSILModule(SILModule *M);
+void visitSILModule(SILModule *M);
```
Comment (karimhamdanali) [inline include/swift/WALASupport/SILWalaInstructionVisitor.h]: please clean up the spacing here

**Review by karimhamdanali — CHANGES_REQUESTED** [review]

**Comment by taufique71** [comment]
@karimhamdanali spacing is cleaned up.

**Comment by karimhamdanali** [comment]
@mmroz I see that @taufique71  has requested your review on this one. Do you approve?

**Comment by mmroz** [comment]
I don’t think this is correct because the instruction is about indexing into an array however the Node created is an ARRAY_REF which would be a mutable array. This instruction does not create an array or even use one

**Comment by taufique71** [comment]
@mmroz In your opinion, from available node types which node type resembles best what this instruction is doing?

**Comment by mmroz** [comment]
Well with the given description of: “Given an address that references into an array of values, returns the address of the %1-th element relative to %0” and knowing %1 is an integer literal and %0 is the size of each element and the start of the array it may be possible to find the element and assign it to the node map

**Comment by taufique71** [comment]
@mmroz can you explain this part a little more?

> %0 is the size of each element and the start of the array

If I know the size of each element and I know the address of the first element then I can get address of n_th element. In that case what would the CAst node type be? But I am confused about size. Do we really know the size of the each element?

Plus I am confused about what ARRAY_REF actually means as there is no description there. Is it possible to clear me up if you know it?

**Comment by mmroz** [comment]
can you review this and see what you think?

```
jobject SILWalaInstructionVisitor::visitIndexAddrInst(IndexAddrInst *IAI) {

  SILValue BaseValue = IAI->getBase();
  SILValue IndexValue = IAI->getIndex();

  jobject IndexNode = findAndRemoveCAstNode(IndexValue.getOpaqueValue());
  jobject BaseNode = findAndRemoveCAstNode(base.getOpaqueValue());

  if (Print) {
    llvm::outs() << "\t\t [BASE ADDR]:" << BaseValue.getOpaqueValue() << "\n";
    llvm::outs() << "\t\t [INDEX ADDR]:" << IndexValue.getOpaqueValue() << "\n";
  }

  jobject IndexVarName = Wala->makeConstant("array_index");
  jobject IndexVar = Wala->makeNode(CAstWrapper::VAR, IndexVarName);
  jobject IndexNode = Wala->makeNode(CAstWrapper::ASSIGN, IndexVar, indexNode);

  auto IndexAddrNode = Wala->makeNode(CAstWrapper::BLOCK_EXPR, IndexNode, BaseNode);

  NodeMap.insert(std::make_pair(static_cast<ValueBase *>(IAI), IndexAddrNode));

  return IndexAddrNode;
}
```

**Comment by taufique71** [comment]
I have a couple of questions on above implementation -

1. Why do we need to create a new `IndexNode` with all those constant node creation, variable node creation and assignment node creation when we know that we can find this IndexNode created somewhere before?

2. `BLOCK_EXPR` resembles a sequential chain of simple expressions right? What I didn't get is how a BLOCK_EXPR resembles specifically array indexing here?

**Comment by mmroz** [comment]
1. The Node we have is an IntegerLiteral but I want to represent an iterator for that value so I created a variable to represent that iterator. Its based off of how this handled memory accessing: https://www.programcreek.com/java-api-examples/index.php?source_dir=wala-mirror-master/com.ibm.wala.cast.js.rhino/source/com/ibm/wala/cast/js/translator/RhinoToAstTranslator.java

2. A BOCK_EXPR in this case first represents iterating (using the above variable) then the result which is the base.

I am not 100% on this but it appears to work!

**Comment by taufique71** [comment]
Thanks for the clarification. I understood your first point but still the second one is not clear to me. I don't know if this type of conversation(due to my lack of understanding) is appropriate here or not. May be real-time communication can help better. Would set up a meeting with you after communicating over Slack. Would update the PR after being clear about what I am doing.

**Review by mmroz — APPROVED** [review]
Looks good!

**Comment by karimhamdanali** [comment]
Have you guys already discussed the final status of this PR? Is it now ready to go?

**Comment by taufique71** [comment]
@karimhamdanali I guess it's ready to go. Mark has also approved the changes after our discussion with you and Julian.

**Review by karimhamdanali — APPROVED** [review]

**Pull request merged**